syntax = "proto2";

package tutorial;

service Greeter {
  // get information of one person
  rpc GetPerson (PersonRequest) returns (Person) {}

  // get information of a list of persons
  rpc ListPersons(PersonRequest) returns (stream Person) {}

  // sent persons info and store remotely
  rpc recordPerson(stream storeInfo) returns (storeConfirmation) {}
}

message PersonRequest {
  required string name = 1;
}

message Person {
  optional string name = 1;
  optional int32 id = 2;
  optional string email = 3;

  enum PhoneType {
    MOBILE = 0;
    HOME = 1;
    WORK = 2;
  }

  message PhoneNumber {
    optional string number = 1;
    optional PhoneType type = 2 [default = HOME];
  }

  repeated PhoneNumber phones = 4;
}

message storeConfirmation {
  required string info = 2;
}

message storeInfo {
  required string file_name = 1;
  required Person psn = 2;
}

message AddressBook {
  repeated Person people = 1;
}